array.go

```diff
@@ -2,43 +2,47 @@ package amf0
 
 import (
 	"bytes"
+	"encoding/binary"
 	"errors"
 	"io"
+	"reflect"
 )
 
 type Array struct {
-	*paired
+	*Paired
 }
 
-var _ AmfType = &Array{}
-var _ Paired = &Array{}
+var _ AmfType = new(Array)
 
 func NewArray() *Array {
-	return &Array{newPaired()}
+	return &Array{NewPaired()}
 }
 
+// Implements AmfType.Marker
+func (a *Array) Marker() byte { return 0x08 }
+
+func (a *Array) Native() reflect.Type { return reflect.TypeOf(a).Elem() }
+
 // Implements AmfType.Decode
 func (a *Array) Decode(r io.Reader) error {
-	numBytes, err := readBytes(r, 4)
-	if err != nil {
+	var n [4]byte
+	if _, err := r.Read(n[:]); err != nil {
```

Comment:
Good catch! Probably want to use [`io.ReadFull`](https://golang.org/pkg/io/#ReadFull) instead, since it is equivalent to calling:

```go
import "io"

func Foo() {
        var but [4]byte

        _, err := io.ReadAtLeast(r, buf[:], len(buf))

        // The same as:
        _, err := io.ReadFull(r, buf[:])
}
```

Will make sure to update those throughout this PR.